### PR TITLE
Add Telegram support, colors and puzzle solving timestamps

### PR DESCRIPTION
**Pull request by iburakov**
Hi there! Thanks for sharing this code, it was very helpful. 

I've done some further work on it for my own requirements and created a [fork](https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram) as a separate repo with its own builds for the time being. 

Just suggesting you to take a look, because I have nothing against merging this work back to yours (if you consider it reasonable, of course).

I've simplified writing custom webhook senders and implemented one for Telegram. I've also added colors and event times to notifications, so now they look like this:

The formatting part is pretty opinionated, yeah. If you don't like it, a way to please everyone would be to accept pattern str in another env var and do a `.format(day=..., part=..., part_emoji=...)`. So, your format becomes `NOTIFICATION_PATTERN="{member} solved day {day} part {part}"`, mine becomes `NOTIFICATION_PATTERN="Day {day}: {member} got {part_emoji} at {when}"`, users can come up with every other thing they like, etc.

So, if you're open to merging it, let me know, because (at least) a README needs some changes for it to be mergeable :) Otherwise feel free to close this PR.

### CODE REVIEW

**Comment by tofran** [comment]
@iburakov Thank you very much for all the effort! I love it, and thank you for preserving the code-style.

I will have a better look later. But the only thing I will point outright now is the emoji, the colored circle is not very representative - at least in my opinion. I would change it to one of the following (1st, 2nd puzzle):

- ⭐, 🌟
- 🌲, 🎄
- 🌱, 🌲/🎄

And yes, I actually like the `NOTIFICATION_PATTERN` idea. It actually allows for easy i18n that would be useful in my case.

Edit: maybe having two patterns, one for each part will solve everything. It can include the emoji in it.

**Comment by tofran** [comment]
Woops the action failed because it is directly pushing the image, I would imagine it would do some sort of linting. Please ignore it. I will fix it in another PR

File: webhook_senders.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import os
+from abc import ABC, abstractmethod
+from venv import logger
+
+import requests
+
+
+class WebhookSender(ABC):
```
Comment (tofran) [inline webhook_senders.py]: I would call this `BaseNotificationSender`.

File: webhook_senders.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    # __init__() should load required parameters from env vars
+
+    @abstractmethod
+    def send(self, content):
+        ...
+
+
+class DefaultSender(WebhookSender):
```
Comment (tofran) [inline webhook_senders.py]: .... and this `WebhookSender`

File: README.md [hunk 3]
```diff
@@ -1,40 +1,68 @@
-# Advent of code leaderboard notifier
+# Advent of Code leaderboard notifier (Telegram-enabled fork)
 
 Send a webhook notification when someone from an Advent Of Code leaderboard solves a puzzle.
 
-I'm using this to track the progress of a group of friends on Discord.
+This is a fork of [tofran/advent-of-code-leaderboard-notifier](https://github.com/tofran/advent-of-code-leaderboard-notifier) that:
```
Comment (tofran) [inline README.md]: As you mentioned this needs work, if you wish, I can help with that. Feel free to enable author contribution permission.

File: README.md [hunk 4]
```diff
@@ -1,40 +1,68 @@
...
+- `WEBHOOK_SENDER`: Optional, allows to select a non-default webhook sender. Can be set to `telegram` to send notifications to Telegram instead. Each webhook sender has its own params. See below and in `webhook_senders.py` for more info.
+
+`default` webhook sender params:
+
+- `WEBHOOK_URL`: Where to send the webhook. For example, a Discord webhook URL.
+
+`telegram` webhook sender params:
 
+- `WEBHOOK_TELEGRAM_URL`: The URL of your Telegram bot's webhook. Usually it's `https://api.telegram.org/bot<token>/sendMessage`. Ask [BotFather](https://t.me/botfather) for a token.
+- `WEBHOOK_TELEGRAM_CHAT_ID`: comma-separated list of chat IDs to send notifications to. See [Telegram docs](https://core.telegram.org/bots/api#sendmessage) for more info. You can find them using bots like [@JsonDumpBot](https://t.me/JsonDumpBot).
```
Comment (tofran) [inline README.md]: Please prefix it with an `S` for plural: `WEBHOOK_TELEGRAM_CHAT_IDS`

**Review by tofran — CHANGES_REQUESTED** [review]
I like it! Thank you for the logging addition.

What do you think of these changes?:

**Comment by iburakov** [comment]
That's nice to hear you like it, thanks! Great idea about distinct patterns for both notifications, that'll be much more flexible, indeed.

Here are my plans for the next few hours:
- address requested changes
- implement getting two notification patterns from the env
- ensure the code passes the linting you've pushed recently

I'll write here when I'm done. You should have the edit rights, feel free to change anything for the merge. ~Please, use the new [to-be-merged-back](https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram/tree/to-be-merged-back) branch.~ Nevermind, use `main` :) Apparently I can't change the source branch of this PR, so we're stuck on main here.

File: webhook_senders.py (hunk 1 above)
Comment (iburakov) [reply webhook_senders.py]: Yeah, that generalizes the concept for even more applications. Ok, let's do it, and I'll try to make sense of everything related like renaming the `WEBHOOK_SENDER` env var to `NOTIFICATION_SENDER` as well.

File: webhook_senders.py (hunk 2 above)
Comment (iburakov) [reply webhook_senders.py]: yep, makes sense, given the change above

File: README.md (hunk 3 above)
Comment (iburakov) [reply README.md]: yeah, you should have the rights, go ahead :)

~use the [to-be-merged-back](https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram/tree/to-be-merged-back) branch~ 
nope, use `main` :)

File: README.md (hunk 4 above)
Comment (iburakov) [reply README.md]: right, thanks

**Comment by iburakov** [comment]
So, after considering a few ways to enable customizing notification patterns and emojis, that's what I suggest as a flexible-yet-convenient way: 

```python
NOTIFICATION_PATTERN_EMOJIS = os.getenv("NOTIFICATION_PATTERN_EMOJIS", "🌱🎄")
NOTIFICATION_PATTERN = os.getenv(
    "NOTIFICATION_PATTERN", "Day {day}: {member} got {part_emoji} at {when}"
)
NOTIFICATION_2_PATTERN = os.getenv("NOTIFICATION_2_PATTERN", NOTIFICATION_PATTERN)
```

It seems to cover basically all discussed needs with minimal effort. Override only `NOTIFICATION_PATTERN_EMOJIS` if you need to change just the emojis, override the whole `NOTIFICATION_PATTERN` if you need to translate it, and, optionally, override `NOTIFICATION_2_PATTERN` if it must be structurally different from the first one.

**Comment by iburakov** [comment]
A friend of mine also suggested to display the time as "[d] hh:mm:ss" since moment of corresponding puzzle publication. I tried it, it's arguably more convenient than absolute :mm:ss.

```
Day 4: Alice got 🌱 after 00:11:16
Day 4: Alice got 🎄 after 00:24:34
Day 4: Bob got 🌱 after 00:58:26
Day 4: Bob got 🎄 after 01:19:11
```

Implemented it as an `after` notification pattern key. How that sounds?  FYI, I've updated the default `NOTIFICATION_PATTERN` to this.

**Comment by iburakov** [comment]
Done, fixed the lint CI workflow as well. Treat all changes here as suggestions that seemed reasonable to me.

Leaving everything up to you now. You should have the edit rights, so change anything as you wish, especially README.

**Review by tofran — APPROVED** [review]
I think it looks great. Thank you for all the work.
Also, sorry for the broken github workflow. Should be working now (at least green on this side).

Can you spot anything or can I merge?

**Comment by iburakov** [comment]
LGTM! Glad this worked.

**Comment by tofran** [comment]
@iburakov right now should be available at `ghcr.io/tofran/advent-of-code-leaderboard-notifier:main`
I will do a release later.

**Comment by iburakov** [comment]
Good, thanks. Also spotted a couple of typos at the second glance (a bit late though, sorry)

---
https://github.com/tofran/advent-of-code-leaderboard-notifier/blob/e3cf8204e2b3ac1a7c9aebaabfb9050ef9e495d1/README.md?plain=1#L86
2dn -> 2nd

---
https://github.com/tofran/advent-of-code-leaderboard-notifier/blob/e3cf8204e2b3ac1a7c9aebaabfb9050ef9e495d1/README.md?plain=1#L5
"group of friends, or co-workers" - extra comma, perhaps

**Comment by tofran** [comment]
Thank you. Fixed those and released the new version: [v1.3.0](https://github.com/tofran/advent-of-code-leaderboard-notifier/releases/tag/v1.3.0)

**Pull request merged**